File: Downloads/GameTech-Store-main - copia/static/js/hardware_analyzer.js
/**
 * Hardware Analyzer - Frontend Logic
 * Maneja el análisis de hardware y visualización de resultados
 */

// Estado global
let currentAnalysis = null;

// Inicializar cuando el DOM esté listo
document.addEventListener('DOMContentLoaded', function() {
    const analyzeBtn = document.getElementById('analyze-btn');
    if (analyzeBtn) {
        analyzeBtn.addEventListener('click', analyzeHardware);
    }
});

/**
 * Función principal de análisis
 */
async function analyzeHardware() {
    const cpuId = document.getElementById('cpu-select').value;
    const gpuId = document.getElementById('gpu-select').value;
    const ramId = document.getElementById('ram-select').value;
    
    // Validar selección
    if (!cpuId || !gpuId || !ramId) {
        showAlert('Por favor selecciona todos los componentes (CPU, GPU y RAM)', 'warning');
        return;
    }
    
    // Mostrar loading
    showLoading(true);
    hideResults();
    
    try {
        const response = await fetch('/api/analizar-hardware', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                cpu_id: Number.parseInt(cpuId),
                gpu_id: Number.parseInt(gpuId),
                ram_id: Number.parseInt(ramId)
            })
        });
        
        if (!response.ok) {
            throw new Error('Error en la respuesta del servidor');
        }
        
        const data = await response.json();
        
        if (data.success) {
            currentAnalysis = data;
            displayResults(data);
        } else {
            showAlert(data.error || 'Error al analizar el hardware', 'danger');
        }
        
    } catch (error) {
        console.error('Error:', error);
        showAlert('Error al conectar con el servidor. Por favor intenta de nuevo.', 'danger');
    } finally {
        showLoading(false);
    }
}

/**
 * Mostrar todos los resultados
 */
function displayResults(data) {
    displaySystemScore(data.system_score);
    displayBottlenecks(data.bottlenecks);
    displayRecommendations(data.recommendations);
    displayGameCompatibility(data.games);
    
    // Mostrar contenedor de resultados
    document.getElementById('results-container').classList.remove('d-none');
    
    // Scroll suave a resultados
    document.getElementById('results-container').scrollIntoView({ 
        behavior: 'smooth', 
        block: 'start' 
    });
}

/**
 * Mostrar puntuación del sistema
 */
function displaySystemScore(score) {
    const container = document.getElementById('system-score-card');
    
    const tierColor = getTierColor(score.total);
    
    container.innerHTML = `
        <div class="card-body text-center py-4" style="background: linear-gradient(135deg, ${tierColor}22 0%, ${tierColor}11 100%);">
            <div class="row align-items-center">
                <div class="col-md-4">
                    <div class="score-display">${score.total}</div>
                    <p class="text-muted mb-0">Puntuación Total</p>
                </div>
                <div class="col-md-4">
                    <h4 class="mb-2" style="color: ${tierColor};">
                        <i class="fas fa-trophy me-2"></i>${score.tier}
                    </h4>
                    <p class="text-muted mb-0">Nivel de Rendimiento</p>
                </div>
                <div class="col-md-4">
                    <div class="row g-2">
                        <div class="col-12">
                            <small class="text-muted">CPU:</small>
                            <strong class="ms-2">${score.cpu_score}</strong>
                        </div>
                        <div class="col-12">
                            <small class="text-muted">GPU:</small>
                            <strong class="ms-2">${score.gpu_score}</strong>
                        </div>
                        <div class="col-12">
                            <small class="text-muted">RAM:</small>
                            <strong class="ms-2">${score.ram_gb}GB</strong>
                        </div>
                    </div>
                </div>
            </div>
            <hr class="my-3">
            <div class="row text-start">
                <div class="col-md-4">
                    <small class="text-muted"><i class="fas fa-microchip me-1"></i>CPU:</small>
                    <p class="mb-0">${score.components.cpu}</p>
                </div>
                <div class="col-md-4">
                    <small class="text-muted"><i class="fas fa-video me-1"></i>GPU:</small>
                    <p class="mb-0">${score.components.gpu}</p>
                </div>
                <div class="col-md-4">
                    <small class="text-muted"><i class="fas fa-memory me-1"></i>RAM:</small>
                    <p class="mb-0">${score.components.ram}</p>
                </div>
            </div>
        </div>
    `;
}

/**
 * Mostrar cuellos de botella
 */
function displayBottlenecks(bottlenecks) {
    const container = document.getElementById('bottlenecks-container');
    
    if (!bottlenecks.has_bottleneck) {
        container.innerHTML = `
            <div class="alert alert-success mb-0">
                <h5><i class="fas fa-check-circle me-2"></i>¡Sistema Balanceado!</h5>
                <p class="mb-0">${bottlenecks.description}</p>
            </div>
        `;
        return;
    }
    
    const severityClass = {
        'severe': 'danger',
        'moderate': 'warning',
        'mild': 'info'
    }[bottlenecks.severity] || 'info';
    
    const severityIcon = {
        'severe': 'exclamation-triangle',
        'moderate': 'exclamation-circle',
        'mild': 'info-circle'
    }[bottlenecks.severity] || 'info-circle';
    
    let html = `
        <div class="alert alert-${severityClass}">
            <h5>
                <i class="fas fa-${severityIcon} me-2"></i>
                Cuello de Botella Detectado
                ${bottlenecks.percentage_loss > 0 ? `<span class="badge bg-${severityClass} ms-2">-${bottlenecks.percentage_loss}% rendimiento</span>` : ''}
            </h5>
            <div class="bottleneck-description">${formatMarkdown(bottlenecks.description)}</div>
    `;
    
    if (bottlenecks.recommendations && bottlenecks.recommendations.length > 0) {
        html += `
            <hr>
            <h6><i class="fas fa-tools me-2"></i>Recomendaciones:</h6>
            <ul class="mb-0">
                ${bottlenecks.recommendations.map(rec => `<li>${rec}</li>`).join('')}
            </ul>
        `;
    }
    
    html += `</div>`;
    container.innerHTML = html;
}

/**
 * Mostrar recomendaciones generales
 */
function displayRecommendations(recommendations) {
    const section = document.getElementById('recommendations-section');
    const container = document.getElementById('recommendations-container');
    
    if (!recommendations || recommendations.length === 0) {
        section.classList.add('d-none');
        return;
    }
    
    section.classList.remove('d-none');
    
    const html = `
        <ul class="list-group list-group-flush">
            ${recommendations.map(rec => `
                <li class="list-group-item">
                    <i class="fas fa-arrow-right text-info me-2"></i>${rec}
                </li>
            `).join('')}
        </ul>
    `;
    
    container.innerHTML = html;
}

/**
 * Mostrar compatibilidad con juegos
 */
function displayGameCompatibility(games) {
    // Actualizar contadores
    document.getElementById('ultra-count').textContent = games.can_run_ultra.length;
    document.getElementById('high-count').textContent = games.can_run_high.length;
    document.getElementById('medium-count').textContent = games.can_run_medium.length;
    document.getElementById('low-count').textContent = games.can_run_low.length;
    document.getElementById('cannot-count').textContent = games.cannot_run.length;
    
    // Renderizar cada categoría
    renderGameCategory('ultra-games', games.can_run_ultra, 'ultra', 'success');
    renderGameCategory('high-games', games.can_run_high, 'high', 'primary');
    renderGameCategory('medium-games', games.can_run_medium, 'medium', 'warning');
    renderGameCategory('low-games', games.can_run_low, 'low', 'secondary');
    renderGameCategory('cannot-games', games.cannot_run, 'cannot', 'danger');
}

/**
 * Renderizar categoría de juegos
 */
function renderGameCategory(containerId, games, quality, badgeColor) {
    const container = document.getElementById(containerId);
    
    if (games.length === 0) {
        container.innerHTML = `
            <div class="text-center text-muted py-4">
                <i class="fas fa-inbox fa-3x mb-3"></i>
                <p>No hay juegos en esta categoría</p>
            </div>
        `;
        return;
    }
    
    const html = games.map(game => `
        <div class="card game-card ${quality} mb-3">
            <div class="card-body">
                <div class="row align-items-center">
                    <div class="col-md-2">
                        <img src="${game.imagen || '/static/images/game-placeholder.jpg'}" 
                             class="img-fluid rounded" 
                             alt="${game.nombre}"
                             onerror="this.src='/static/images/game-placeholder.jpg'">
                    </div>
                    <div class="col-md-6">
                        <h5 class="mb-1">${game.nombre}</h5>
                        ${game.reason ? `<small class="text-danger">${game.reason}</small>` : ''}
                        ${game.bottleneck ? `
                            <div class="mt-2">
                                <span class="badge bg-warning text-dark">
                                    <i class="fas fa-exclamation-triangle me-1"></i>
                                    Limitado por ${game.bottleneck.toUpperCase()}
                                </span>
                            </div>
                        ` : ''}
                    </div>
                    <div class="col-md-2 text-center">
                        ${game.expected_fps > 0 ? `
                            <div class="fps-indicator text-${badgeColor}">
                                ${game.expected_fps} FPS
                            </div>
                            <small class="text-muted">Estimado</small>
                        ` : '<small class="text-danger">No puede correr</small>'}
                    </div>
                    <div class="col-md-2 text-center">
                        <span class="badge bg-${badgeColor} quality-badge">
                            ${getQualityLabel(quality)}
                        </span>
                        <div class="mt-2">
                            <small class="text-muted">$${game.precio}</small>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    `).join('');
    
    container.innerHTML = html;
}

/**
 * Utilidades
 */

function getTierColor(score) {
    if (score >= 15000) return '#9c27b0';
    if (score >= 10000) return '#2196f3';
    if (score >= 7000) return '#4caf50';
    if (score >= 4000) return '#ff9800';
    return '#f44336';
}

function getQualityLabel(quality) {
    const labels = {
        'ultra': 'Ultra (4K)',
        'high': 'Alto (1440p)',
        'medium': 'Medio (1080p)',
        'low': 'Bajo (720p)',
        'cannot': 'No Compatible'
    };
    return labels[quality] || quality;
}

function formatMarkdown(text) {
    // Convertir markdown simple a HTML
    return text
        .replaceAll(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
        .replaceAll(/\n\n/, '</p><p>')
        .replaceAll(/\n/, '<br>');
}

function showLoading(show) {
    const loader = document.getElementById('loading-container');
    if (show) {
        loader.classList.remove('d-none');
    } else {
        loader.classList.add('d-none');
    }
}

function hideResults() {
    document.getElementById('results-container').classList.add('d-none');
}

function showAlert(message, type = 'info') {
    // Crear alert temporal
    const alertDiv = document.createElement('div');
    alertDiv.className = `alert alert-${type} alert-dismissible fade show position-fixed top-0 start-50 translate-middle-x mt-3`;
    alertDiv.style.zIndex = '9999';
    alertDiv.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
    
    document.body.appendChild(alertDiv);
    
    // Auto-remover después de 5 segundos
    setTimeout(() => {
        alertDiv.remove();
    }, 5000);
}

File: Downloads/GameTech-Store-main - copia/controllers/hardware_analyzer.py
"""
Controlador para el analizador de hardware
Permite a los usuarios analizar su configuración y ver compatibilidad con juegos
"""
from flask import Blueprint, render_template, request, jsonify
from models.database_models import Hardware, Game, GameRequirements
from utils.bottleneck_detector import BottleneckDetector
from utils.performance_calculator import PerformanceCalculator

analyzer_bp = Blueprint('analyzer', __name__)

@analyzer_bp.route('/analizador-hardware')
def hardware_analyzer_page():
    """Página principal del analizador de hardware"""
    # Cargar componentes para los selectores
    cpus = Hardware.get_hardware_by_tipo('CPU')
    gpus = Hardware.get_hardware_by_tipo('GPU')
    rams = Hardware.get_hardware_by_tipo('RAM')
    
    return render_template('hardware_checker.html',
                         cpus=cpus,
                         gpus=gpus,
                         rams=rams)

@analyzer_bp.route('/api/analizar-hardware', methods=['POST'])
def analyze_hardware():
    """API para analizar la configuración del usuario"""
    try:
        data = request.get_json()
        
        cpu_id = data.get('cpu_id')
        gpu_id = data.get('gpu_id')
        ram_id = data.get('ram_id')
        
        # Validar datos
        if not all([cpu_id, gpu_id, ram_id]):
            return jsonify({'error': 'Faltan componentes'}), 400
        
        # Obtener componentes
        cpu = Hardware.get_hardware_by_id(cpu_id)
        gpu = Hardware.get_hardware_by_id(gpu_id)
        ram = Hardware.get_hardware_by_id(ram_id)
        
        if not all([cpu, gpu, ram]):
            return jsonify({'error': 'Componentes no encontrados'}), 404
        
        # 1. Calcular puntuación del sistema
        system_score = calculate_system_score(cpu, gpu, ram)
        
        # 2. Detectar cuellos de botella
        bottlenecks = BottleneckDetector.detect(cpu, gpu, ram)
        
        # 3. Analizar compatibilidad con juegos
        games_analysis = analyze_game_compatibility(cpu, gpu, ram)
        
        # 4. Generar recomendaciones
        recommendations = generate_recommendations(bottlenecks, system_score)
        
        return jsonify({
            'success': True,
            'system_score': system_score,
            'bottlenecks': bottlenecks,
            'games': games_analysis,
            'recommendations': recommendations
        })
    
    except Exception as e:
        return jsonify({'error': str(e)}), 500

def calculate_system_score(cpu, gpu, ram):
    """Calcular puntuación general del sistema"""
    cpu_score = cpu.benchmark_score or 0
    gpu_score = gpu.benchmark_score or 0
    ram_gb = ram.get_ram_capacity_gb() if hasattr(ram, 'get_ram_capacity_gb') else 8
    ram_score = ram_gb * 100  # 16GB = 1600 puntos
    
    # Ponderación: GPU 50%, CPU 35%, RAM 15%
    total = (gpu_score * 0.5) + (cpu_score * 0.35) + (ram_score * 0.15)
    
    return {
        'total': int(total),
        'cpu_score': cpu_score,
        'gpu_score': gpu_score,
        'ram_score': ram_score,
        'ram_gb': ram_gb,
        'tier': get_performance_tier(total),
        'components': {
            'cpu': f'{cpu.marca} {cpu.modelo}',
            'gpu': f'{gpu.marca} {gpu.modelo}',
            'ram': f'{ram_gb}GB RAM'
        }
    }

def get_performance_tier(score):
    """Determinar el nivel de rendimiento"""
    if score >= 15000:
        return 'Ultra High-End (4K Ultra)'
    elif score >= 10000:
        return 'High-End (1440p Ultra)'
    elif score >= 7000:
        return 'Mid-High (1080p Ultra)'
    elif score >= 4000:
        return 'Mid-Range (1080p Medium-High)'
    else:
        return 'Entry Level (1080p Low-Medium)'

def analyze_game_compatibility(cpu, gpu, ram):
    """Analizar qué juegos puede correr el usuario"""
    games = Game.get_all_games()
    
    results = {
        'can_run_ultra': [],
        'can_run_high': [],
        'can_run_medium': [],
        'can_run_low': [],
        'cannot_run': []
    }
    
    for game in games:
        requirements = GameRequirements.get_by_game_id(game.id)
        
        if not requirements:
            continue  # Skip juegos sin requisitos
        
        # Calcular rendimiento
        performance = PerformanceCalculator.calculate_game_performance(
            cpu, gpu, ram, requirements
        )
        
        game_data = {
            'id': game.id,
            'nombre': game.nombre,
            'imagen': game.imagen,
            'precio': game.precio,
            'expected_fps': performance['fps_estimate'],
            'quality': performance['quality'],
            'bottleneck': performance['bottleneck'],
            'reason': performance.get('reason', '')
        }
        
        # Clasificar por calidad
        if not performance['can_run']:
            results['cannot_run'].append(game_data)
        elif performance['quality'] == 'ultra':
            results['can_run_ultra'].append(game_data)
        elif performance['quality'] == 'high':
            results['can_run_high'].append(game_data)
        elif performance['quality'] == 'medium':
            results['can_run_medium'].append(game_data)
        else:  # low
            results['can_run_low'].append(game_data)
    
    return results

def generate_recommendations(bottlenecks, system_score):
    """Generar recomendaciones de mejora"""
    recommendations = []
    
    # Recomendaciones de bottleneck
    if bottlenecks['has_bottleneck']:
        recommendations.extend(bottlenecks['recommendations'])
    
    # Recomendaciones por tier
    total_score = system_score['total']
    
    if total_score < 7000:
        recommendations.append(
            '💡 Tu sistema es entry-level. Considera actualizar GPU y CPU para mejor experiencia.'
        )
    elif total_score < 10000:
        recommendations.append(
            '💡 Tu sistema es mid-range. Una GPU mejor te daría un salto significativo en rendimiento.'
        )
    
    # Recomendación de RAM
    if system_score['ram_gb'] < 16:
        recommendations.append(
            '💾 16GB de RAM es el estándar actual para gaming. Considera expandir.'
        )
    
    return recommendations
